### ISSUE TITLE
New Algorithm (RACF-KDFAES)

### ISSUE THREAD

**Issue by bigendiansmalls** [hashcat/hashcat#2299]
Hey All -

Want to get the KDFAES algorithm for RACF (z/OS Mainframe) added. I've added it to JTR some time ago and it's pretty stable now.  
(original commit below, was tweaked for performance etc since)

[Original JTR Commit](https://github.com/magnumripper/JohnTheRipper/commit/bdf40440cb33398f07a99509ebeee32699d039c7)

There are test hashes / strings in the commit and I'm available for any further help or testing you might want. I have a ton of info on the constraints / working so let me know what you need.

*example hash*
```CHAD1:$racf$*CHAD1*E7D7E66D00018000000800320010001092BFA567A2FC3FC5B8583F265370CCE01C78FC8CE54A0F2D2836367BDA6E8403```
```P@ssw0rd         (CHAD1)```

Thanks!

-BES

**Comment by jsteube** [comment]
We are working on a tutorial on how to add new plugins to hashcat. Would you be willing to try to add it yourself based on the tutorial we'll give you all the support from hashcat development side?

**Comment by bigendiansmalls** [comment]
I'm willing to give it a try!

**Comment by jsteube** [comment]
Cool, I'll ping you when the tutorial is ready.

**Comment by Viss** [comment]
HI!
PLUS ONEING THIS BECAUSE BIGGIE IS MY HERO :D :D :D

**Comment by bigendiansmalls** [comment]
Hey so this didn't work out so well; no one's pinged me yet and I don't now have time to do it myself.  what's the next best option? @jsteube Cheers :)

**Comment by bigendiansmalls** [comment]
Hey all!  I'd love to get support for this in hashcat if someone wants to point me in the right direction?

**Comment by jsteube** [comment]
The tutorial is ready since a while: https://github.com/hashcat/hashcat/blob/master/docs/hashcat-plugin-development-guide.md

Also, looks like someone implemented it: https://github.com/hashcat/hashcat/pull/2777

So far I had no time to check it, but I'll probably at a later time.

**Comment by bigendiansmalls** [comment]
Fantastic. I’ll review it. Thank you

On Apr 29, 2021, at 09:06, Jens Steube ***@***.***> wrote:

﻿

The tutorial is ready since a while: https://github.com/hashcat/hashcat/blob/master/docs/hashcat-plugin-development-guide.md

Also, looks like someone implemented it: #2777<https://github.com/hashcat/hashcat/pull/2777>

So far I had no time to check it, but I'll probably at a later time.

**Comment by ecostin** [comment]
Chad, it is the same code you have on the portable EPAS unit, just a bit cleaner and a bit optimized. Since we are limited to 8 chars, we did not implement mode 1. Cheers!

**Comment by bigendiansmalls** [comment]
yeah fantastic !

On Apr 29, 2021, at 09:16, ecostin ***@***.***> wrote:

﻿

Chad, it is the same code you have on the portable EPAS unit, just a bit cleaner and a bit optimized. Since we are limited to 8 chars, we did not implement mode 1. Cheers!

**Comment by lnlyssg** [comment]
@bigendiansmalls or @ecostin do you have any plans to take another look at this? I can see the pull request was closed in early 2021 due to the plugin interface changes but that looks to have been the last activity….

**Comment by ecostin** [comment]
Good that you reminded me, will try to have it fixed over the next few weeks.

**Comment by lnlyssg** [comment]
Time for my annual query to see if there's any news :)

**Comment by ecostin** [comment]
And for the usual reply of, God, how quickly that year has passed! And we had no time to clean up the code properly to fit the quality standards in hashcat - no pun here. In the meantime, if you just need it to work, grab it from here: [https://github.com/ecostin/hashcat](https://github.com/ecostin/hashcat)

That clone should be almost synced with hashcat and has a few extra algos:
Hash-Mode 5710 (Cisco-ISE Hashed Password (SHA256))
Hash-Mode 8501 (IBM RACF KDFAES)
Hash-Mode 8502 (IBM AS400 DES)
Hash-Mode 8503 (IBM AS400 SHA1)

Contributors who may want to clean-up our code to get it into the mainstream are welcome.

**Comment by matrix** [comment]
> And for the usual reply of, God, how quickly that year has passed! And we had no time to clean up the code properly to fit the quality standards in hashcat - no pun here. In the meantime, if you just need it to work, grab it from here: https://github.com/ecostin/hashcat
> 
> That clone should be almost synced with hashcat and has a few extra algos: Hash-Mode 5710 (Cisco-ISE Hashed Password (SHA256)) Hash-Mode 8501 (IBM RACF KDFAES) Hash-Mode 8502 (IBM AS400 DES) Hash-Mode 8503 (IBM AS400 SHA1)
> 
> Contributors who may want to clean-up our code to get it into the mainstream are welcome.

hi @ecostin, i'm working on this. Where you found the test vectors for each new hash-mode? Thanks

**Comment by ecostin** [comment]
> > And for the usual reply of, God, how quickly that year has passed! And we had no time to clean up the code properly to fit the quality standards in hashcat - no pun here. In the meantime, if you just need it to work, grab it from here: https://github.com/ecostin/hashcat
> > That clone should be almost synced with hashcat and has a few extra algos: Hash-Mode 5710 (Cisco-ISE Hashed Password (SHA256)) Hash-Mode 8501 (IBM RACF KDFAES) Hash-Mode 8502 (IBM AS400 DES) Hash-Mode 8503 (IBM AS400 SHA1)
> > Contributors who may want to clean-up our code to get it into the mainstream are welcome.
> 
> hi [@ecostin](https://github.com/ecostin), i'm working on this. Where you found the test vectors for each new hash-mode? Thanks

Hi @matrix, great, thank you! The test vectors were collected from the real targets, e.g., mainframe, as400, Cisco. If you are working on this, there is already another module for IBM RACF KDFAES in hashcat, neatly written, but sadly much slower (say half the speed or lower) than ours - maybe you can have a look into that too if you are so generous :) Cheers!

**Comment by matrix** [comment]
> > > And for the usual reply of, God, how quickly that year has passed! And we had no time to clean up the code properly to fit the quality standards in hashcat - no pun here. In the meantime, if you just need it to work, grab it from here: https://github.com/ecostin/hashcat
> > > That clone should be almost synced with hashcat and has a few extra algos: Hash-Mode 5710 (Cisco-ISE Hashed Password (SHA256)) Hash-Mode 8501 (IBM RACF KDFAES) Hash-Mode 8502 (IBM AS400 DES) Hash-Mode 8503 (IBM AS400 SHA1)
> > > Contributors who may want to clean-up our code to get it into the mainstream are welcome.
> > 
> > 
> > hi [@ecostin](https://github.com/ecostin), i'm working on this. Where you found the test vectors for each new hash-mode? Thanks
> 
> Hi [@matrix](https://github.com/matrix), great, thank you! The test vectors were collected from the real targets, e.g., mainframe, as400, Cisco. If you are working on this, there is already another module for IBM RACF KDFAES in hashcat, neatly written, but sadly much slower (say half the speed or lower) than ours - maybe you can have a look into that too if you are so generous :) Cheers!

Hi @ecostin, how did you compare performance, and what device did you use? I'm trying with the benchmark, but it seems like the computation never ends when using your repository, specifically module 8501. Thanks!

**Comment by matrix** [comment]
with NVIDIA GTX 4080

using your

```
------------------------------
* Hash-Mode 8501 (RACF KDFAES)
------------------------------

Speed.#01........:      774 H/s (97.55ms) @ Accel:1 Loops:1 Thr:1 Vec:1

Started: Mon Aug  4 09:08:47 2025
Stopped: Mon Aug  4 09:20:42 2025
```

using the one into hashcat

```
-----------------------------------------------
* Hash-Mode 14200 (RACF KDFAES) [Iterations: 8]
-----------------------------------------------

Speed.#01........:    85223 H/s (0.05ms) @ Accel:4 Loops:1 Thr:256 Vec:1

Started: Mon Aug  4 09:25:00 2025
Stopped: Mon Aug  4 09:25:14 2025
```

let me know

**Comment by ecostin** [comment]
Yeah, the auto tuning is fritzed. Just run it by setting the kernel parameters manually. See below with NVIDIA 3090:

v7.0.0-23-gfbcba66
run with default options
```
-----------------------------------------------
* Hash-Mode 14200 (RACF KDFAES) [Iterations: 8]
-----------------------------------------------

Speed.#01........:    43068 H/s (0.04ms) @ Accel:6 Loops:1 Thr:64 Vec:1

Started: Mon Aug  4 09:52:50 2025
Stopped: Mon Aug  4 09:53:20 2025
```
v6.2.6
run with same kernel parameters as 14200:
./hashcat -b -m8501 -w 4 --kernel-accel=8 --kernel-loops=1 --kernel-threads=64 --force 
```
------------------------------
* Hash-Mode 8501 (RACF KDFAES)
------------------------------

Speed.#01........:    74435 H/s (563.53ms) @ Accel:8 Loops:1 Thr:64 Vec:1

Started: Mon Aug  4 10:04:59 2025
Stopped: Mon Aug  4 10:05:10 2025
```
v7.0.0-23-gfbcba66
run with same kernel parameters as 14200:
./hashcat -b -m8501 -w 4 --kernel-accel=8 --kernel-loops=1 --kernel-threads=64 --force 
```
------------------------------
* Hash-Mode 8501 (RACF KDFAES)
------------------------------

Speed.#01........:    63592 H/s (659.68ms) @ Accel:8 Loops:1 Thr:64 Vec:1

Started: Mon Aug  4 10:04:36 2025
Stopped: Mon Aug  4 10:04:48 2025
```

**Comment by matrix** [comment]
> Yeah, the auto tuning is fritzed. Just run it by setting the kernel parameters manually. See below with NVIDIA 3090:
> 
> v7.0.0-23-gfbcba66 run with default options
> 
> ```
> -----------------------------------------------
> * Hash-Mode 14200 (RACF KDFAES) [Iterations: 8]
> -----------------------------------------------
> 
> Speed.#01........:    43068 H/s (0.04ms) @ Accel:6 Loops:1 Thr:64 Vec:1
> 
> Started: Mon Aug  4 09:52:50 2025
> Stopped: Mon Aug  4 09:53:20 2025
> ```
> 
> v6.2.6 run with same kernel parameters as 14200: ./hashcat -b -m8501 -w 4 --kernel-accel=8 --kernel-loops=1 --kernel-threads=64 --force
> 
> ```
> ------------------------------
> * Hash-Mode 8501 (RACF KDFAES)
> ------------------------------
> 
> Speed.#01........:    74435 H/s (563.53ms) @ Accel:8 Loops:1 Thr:64 Vec:1
> 
> Started: Mon Aug  4 10:04:59 2025
> Stopped: Mon Aug  4 10:05:10 2025
> ```
> 
> v7.0.0-23-gfbcba66 run with same kernel parameters as 14200: ./hashcat -b -m8501 -w 4 --kernel-accel=8 --kernel-loops=1 --kernel-threads=64 --force
> 
> ```
> ------------------------------
> * Hash-Mode 8501 (RACF KDFAES)
> ------------------------------
> 
> Speed.#01........:    63592 H/s (659.68ms) @ Accel:8 Loops:1 Thr:64 Vec:1
> 
> Started: Mon Aug  4 10:04:36 2025
> Stopped: Mon Aug  4 10:04:48 2025
> ```

If your card performs better with certain parameters, you can use tuning files to map them. Otherwise, I wouldn't change the current implementation at the moment. We'll definitely take a look to see if there are any areas that can be optimized.

That said, thank you for providing us with the basis on which to work to officially support them. Feel free to make a PR to update the credits.txt file ;)

**Comment by ecostin** [comment]
Right, tested on 4070, the speed is pretty much the same. Probably only affecting the 3090 and some other cards. Nothing urgent. 

Thanks for the tip, will definitely look into the credits file, marketing is always welcome ;)

**Comment by jsteube** [comment]
The problem is because autotuner doesn't check for loop2 kernels. They are super rare. IF you want to get some decent speeds, try this patch:

```
diff --git a/src/modules/module_14200.c b/src/modules/module_14200.c
index f271e95b6..fb2d2158c 100644
--- a/src/modules/module_14200.c
+++ b/src/modules/module_14200.c
@@ -23,6 +23,7 @@ static const u64   KERN_TYPE      = 14200;
 static const u32   OPTI_TYPE      = OPTI_TYPE_ZERO_BYTE;
 static const u64   OPTS_TYPE      = OPTS_TYPE_ST_UPPER
                                   | OPTS_TYPE_HASH_COPY
+                                  | OPTS_TYPE_NATIVE_THREADS
                                   | OPTS_TYPE_INIT2
                                   | OPTS_TYPE_LOOP2;
 static const u32   SALT_TYPE      = SALT_TYPE_EMBEDDED;
@@ -86,6 +87,34 @@ typedef struct racf_kdfaes
 
 static const char *SIGNATURE_RACF_KDFAES = "$racf-kdfaes$";
 
+u32 module_kernel_accel_min (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const user_options_t *user_options, MAYBE_UNUSED const user_options_extra_t *user_options_extra)
+{
+  const u32 kernel_accel_min = 64;
+
+  return kernel_accel_min;
+}
+
+u32 module_kernel_accel_max (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const user_options_t *user_options, MAYBE_UNUSED const user_options_extra_t *user_options_extra)
+{
+  const u32 kernel_accel_max = 128;
+
+  return kernel_accel_max;
+}
+
+u32 module_kernel_loops_min (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const user_options_t *user_options, MAYBE_UNUSED const user_options_extra_t *user_options_extra)
+{
+  const u32 kernel_loops_min = 8;
+
+  return kernel_loops_min;
+}
+
+u32 module_kernel_loops_max (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const user_options_t *user_options, MAYBE_UNUSED const user_options_extra_t *user_options_extra)
+{
+  const u32 kernel_loops_max = 8;
+
+  return kernel_loops_max;
+}
+
 char *module_jit_build_options (MAYBE_UNUSED const hashconfig_t *hashconfig, MAYBE_UNUSED const user_options_t *user_options, MAYBE_UNUSED const user_options_extra_t *user_options_extra, MAYBE_UNUSED const hashes_t *hashes, MAYBE_UNUSED const hc_device_param_t *device_param)
 {
   char *jit_build_options = NULL;
@@ -344,10 +373,10 @@ void module_init (module_ctx_t *module_ctx)
   module_ctx->module_hook_size                = MODULE_DEFAULT;
   module_ctx->module_jit_build_options        = module_jit_build_options;
   module_ctx->module_jit_cache_disable        = MODULE_DEFAULT;
-  module_ctx->module_kernel_accel_max         = MODULE_DEFAULT;
-  module_ctx->module_kernel_accel_min         = MODULE_DEFAULT;
-  module_ctx->module_kernel_loops_max         = MODULE_DEFAULT;
-  module_ctx->module_kernel_loops_min         = MODULE_DEFAULT;
+  module_ctx->module_kernel_accel_max         = module_kernel_accel_max;
+  module_ctx->module_kernel_accel_min         = module_kernel_accel_min;
+  module_ctx->module_kernel_loops_max         = module_kernel_loops_max;
+  module_ctx->module_kernel_loops_min         = module_kernel_loops_min;
   module_ctx->module_kernel_threads_max       = MODULE_DEFAULT;
   module_ctx->module_kernel_threads_min       = MODULE_DEFAULT;
   module_ctx->module_kern_type                = module_kern_type;
```

On my 4090 it improves by a good margin. Not sure about other cards, but runtime is so low, it should have positive impact on all of the cards. No manual patching should be needed afterwards, only if you really want to get everything out of it.

**Comment by ecostin** [comment]
Thanks! This brings the speed of 14200 to the same speed as our old 8501 on all the cards I tested. Although after the patch it sometimes throws on a NVIDIA GeForce RTX 3090, 23848/24135 MB, 82MCU the error below.
```
* Device #1: Not enough allocatable device memory or free host memory for mapping.
```

**Comment by jsteube** [comment]
I just switched from sysinfo() to /proc/meminfo for host memory reading. If you were on Linux, pull latest master. High chance that the issue is already solved. Please send in a PR if you're fine with the patch on speed.

**Comment by ecostin** [comment]
No more errors with the latest master. Created PR with your patch.

> I just switched from sysinfo() to /proc/meminfo for host memory reading. If you were on Linux, pull latest master. High chance that the issue is already solved. Please send in a PR if you're fine with the patch on speed.
